### PR TITLE
Merge V0.2 vis into V0.2

### PR DESCRIPTION
**Pull request by Meroy9819**
# Description
Merge branch v0.2_vis into v0.2

## Linked issue(s)/Pull request(s)

- [issue_number](issue_link)

## Type of Change

- [ ] Non-breaking bug fix
- [ ] Breaking bug fix
- [x] New feature
- [ ] Test
- [ ] Doc update
- [ ] Docker update

## Related Component

- [x] Simulation toolkit
- [ ] RL toolkit
- [ ] Distributed toolkit

## Has Been Tested

- OS:
  - [x] Windows
  - [ ] Mac OS
  - [ ] Linux
- Python version:
  - [x] 3.6
  - [ ] 3.7
- Key information snapshot(s):
  
## Needs Follow Up Actions

- [ ] New release package
- [ ] New docker image

## Checklist

- [x] Add/update the related comments
- [x] Add/update the related tests
- [x] Add/update the related documentations
- [ ] Update the dependent downstream modules usage

### CODE REVIEW

File: maro/cli/inspector/citi_bike_dashboard.py [hunk 1]
```diff
@@ -0,0 +1,333 @@
...
+    -source_path
+        --epoch_0: Data of each epoch.
+            --stations.csv: Record stations' attributes in this file.
+            --matrices.csv: Record transfer volume information in this file.
+            --stations_summary.csv: Record the summary data of current epoch.
+        ………………
+        --epoch_{epoch_num-1}
+        --manifest.yml: Record basic info like scenario name, name of index_name_mapping file.
+        --full_stations.json: Record the relationship between ports' index and name.
+        --stations_summary.csv: Record cross-epoch summary data.
```
Comment (Meroy9819) [inline maro/cli/inspector/citi_bike_dashboard.py:28]: le comment

File: docs/source/key_components/dashboard_visualization.rst [hunk 2]
```diff
@@ -0,0 +1,174 @@
+Dashboard Visualization
+=======================
+
+Env-dashboard is a post-experimental visualization tool, aims to provide
+more intuitive environment information, which will guide the design of
+the algorithm and continually fine-tuning.
+
+Currently, the visualization of senario **Container Inventory Management**
```
Comment (zhawan) [inline docs/source/key_components/dashboard_visualization.rst:8]: senario -> scenarios

**Review by zhawan — APPROVED** [review]

File: tests/communication/test_rejoin.py [hunk 3]
```diff
@@ -9,7 +9,8 @@
 import unittest
 
 from maro.communication import Proxy, SessionMessage, SessionType
-from tests.communication.utils import get_random_port
+from .utils import get_random_port
+
```
Comment (ArthurJiang) [inline tests/communication/test_rejoin.py:13]: why you need change these?

File: requirements.dev.txt [hunk 4]
```diff
@@ -27,10 +27,14 @@ pandas==0.25.3
...
 sphinx_rtd_theme
 jinja2
 flake8
 PuLP=2.1
+streamlit==0.69.1
+altair==4.1.0
+tqdm==4.51.0
 editorconfig-checker
 aria2p==0.9.1
+
```
Comment (ArthurJiang) [inline requirements.dev.txt:40]: rm

File: maro/simulator/scenarios/cim/business_engine.py [hunk 5]
```diff
@@ -98,11 +98,6 @@ def snapshots(self) -> SnapshotList:
         """SnapshotList: Snapshot list of current frame."""
         return self._snapshots
```
Comment (ArthurJiang) [inline maro/simulator/scenarios/cim/business_engine.py:100]: why you rm these?

File: maro/requirements.build.txt [hunk 6]
```diff
@@ -1,2 +1,5 @@
+numpy==1.19.1
 Cython==0.29.14
-numpy==1.19.1
```
Comment (ArthurJiang) [inline maro/requirements.build.txt:2]: why build need these?

File: maro/data_lib/dump_csv_converter.py [hunk 7]
```diff
@@ -61,7 +60,7 @@ def get_new_snapshot_folder(self):
         self._last_snapshot_folder = folder
         return folder
 
-    def process_data(self, filesource: str):
+    def process_data(self, config_data: dict):
```
Comment (ArthurJiang) [inline maro/data_lib/dump_csv_converter.py:63]: what's the meaning of config_data?

File: maro/cli/inspector/citi_bike_dashboard.py [hunk 8]
```diff
@@ -0,0 +1,333 @@
+import os
```
Comment (ArthurJiang) [inline maro/cli/inspector/citi_bike_dashboard.py:1]: add licensed for all files.

File: docs/source/scenarios/container_inventory_management.rst [hunk 9]
```diff
@@ -633,11 +633,79 @@ random mode, we hope this could help you learn how to use the environment interf
 
            # Drive the environment with the random action.
            metrics, decision_event, is_done = env.step(action)
-
+
```
Comment (ArthurJiang) [inline docs/source/scenarios/container_inventory_management.rst:636]: rm space

File: docs/source/key_components/dashboard_visualization.rst [hunk 10]
```diff
@@ -0,0 +1,174 @@
+Dashboard Visualization
```
Comment (ArthurJiang) [inline docs/source/key_components/dashboard_visualization.rst:1]: add this page to index

File: docs/source/key_components/dashboard_visualization.rst [hunk 11]
```diff
@@ -0,0 +1,174 @@
...
+Currently, the visualization of senario **Container Inventory Management**
+and **Citi Bike** are supported.
+
+Dependency
+----------
+
+Module **streamlit** and **altair** should be pre-installed.
+
+* `streamlit <https://www.streamlit.io/>`_: An open-source app framework.
+
```
Comment (ArthurJiang) [inline docs/source/key_components/dashboard_visualization.rst:17]: meaningless

File: docs/source/key_components/dashboard_visualization.rst [hunk 12]
```diff
@@ -0,0 +1,174 @@
...
+
+* `altair <https://altair-viz.github.io/>`_: A declarative statistical visualization library.
+
+install them with:
```
Comment (ArthurJiang) [inline docs/source/key_components/dashboard_visualization.rst:21]: Install ...

File: docs/source/key_components/dashboard_visualization.rst [hunk 13]
```diff
@@ -0,0 +1,174 @@
...
+The command line would print out the selected port.
+
+Feature List
+------------
+Basically, each scenario has 2 parts of visualization: intra-epoch view
+and inter-epoch view. User could switch between them freely.
+
+Intra-epoch view
+~~~~~~~~~~~~~~~~
+
```
Comment (ArthurJiang) [inline docs/source/key_components/dashboard_visualization.rst:98]: rm blank line

File: docs/source/key_components/dashboard_visualization.rst [hunk 14]
```diff
@@ -0,0 +1,174 @@
...
+----
+
+If any file is missed compared with the expected folder structure
+displayed above, the command line would prompt users with an error message.
+The visualization tool looks for the free port to launch page in sequence, starting with port 8501.
+The command line would print out the selected port.
+
+Feature List
+------------
```
Comment (ArthurJiang) [inline docs/source/key_components/dashboard_visualization.rst:92]: Do we need a blank line, here?

File: docs/source/key_components/dashboard_visualization.rst [hunk 15]
```diff
@@ -0,0 +1,174 @@
...
+Detailed introduction please refer to 
+`Citi Bike Visualization <../scenarios/citi_bike.html#Visualization>`_.
+
+Epoch/Snapshot/Resource Holder Index Selection
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+To view the details of a resource holder or a tick, user could select
+the specific index of epoch/snapshot/resource holder by sliding the slider
+on the left side of page.
+
```
Comment (ArthurJiang) [inline docs/source/key_components/dashboard_visualization.rst:126]: we need a figure for this feature?

Comment (ArthurJiang) [reply docs/source/key_components/dashboard_visualization.rst:126]: also for others

File: docs/source/key_components/distributed_toolkit.rst [hunk 16]
```diff
@@ -35,9 +35,9 @@ There are two key components in the distributed toolkit:
   ``(i)send``\ , ``receive``\ , ``(i)broadcast``\ , ``(i)scatter``\ , etc. The communication
   component use a replaceable communication protocol driver to adopt different
   communication protocol stack (e.g. `TCP/IP <https://en.wikipedia.org/wiki/Internet_protocol_suite>`_\ ,
-  `InfiniBand <https://en.wikipedia.org/wiki/InfiniBand>`_). Check the
-  `distributed communication <./communication.html>`_ to get more details.
+  `InfiniBand <https://en.wikipedia.org/wiki/InfiniBand#:~:text=InfiniBand%20(IB>`_\ %20is%20a%20computer,both%20among%20and%20within%20computers.)
```
Comment (ArthurJiang) [inline docs/source/key_components/distributed_toolkit.rst:38]: ?????

File: docs/source/scenarios/container_inventory_management.rst [hunk 17]
```diff
@@ -633,11 +633,79 @@ random mode, we hope this could help you learn how to use the environment interf
...
+    
        # Query for the environment business metrics at the end of each episode,
        # it is usually users' optimized object in CIM scenario (usually includes multi-target).
        print(f"ep: {ep}, environment metrics: {env.metrics}")
        env.reset()
 
 Jump to `this notebook <https://github.com/microsoft/maro/tree/master/notebooks/container_inventory_management/interact_with_environment.ipynb>`_
 for a quick experience.
+
+Visualization
```
Comment (ArthurJiang) [inline docs/source/scenarios/container_inventory_management.rst:645]: mv all vis related before line 642: Jump to ..., also for others

File: maro/simulator/scenarios/cim/business_engine.py (hunk 5 above)
Comment (Meroy9819) [reply maro/simulator/scenarios/cim/business_engine.py:100]: this should be deleted.

File: docs/source/key_components/dashboard_visualization.rst [hunk 18]
```diff
@@ -0,0 +1,186 @@
...
+----
+
+How to Use?
+-----------
+
+Get Data
+~~~~~~~~
+
+Experimental data is dumped automatically while running the experiment.
```
Comment (ArthurJiang) [inline docs/source/key_components/dashboard_visualization.rst:35]: we should have a switch for dump or not

File: docs/source/key_components/dashboard_visualization.rst [hunk 19]
```diff
@@ -0,0 +1,186 @@
...
+User need to run experiment through the file in folder maro/examples in source code. 
+User could specify dump destination folder by setting the parameter **opts** as below:
+
+.. code-block:: sh
+
+    opts['enable-dump-snapshot'] = EXPECTED_OUTPUT_FOLDER
```
Comment (ArthurJiang) [inline docs/source/key_components/dashboard_visualization.rst:41]: we also need code snippets for enabling dump and changing dump target folder.

@Meroy9819 Do you know how to run based on these?

File: tests/cli/inspector/test_dashboard.py [hunk 20]
```diff
@@ -0,0 +1,24 @@
+import unittest
```
Comment (ArthurJiang) [inline tests/cli/inspector/test_dashboard.py:1]: license

File: maro/cli/inspector/launch_env_dashboard.py [hunk 21]
```diff
@@ -0,0 +1,23 @@
+
```
Comment (ArthurJiang) [inline maro/cli/inspector/launch_env_dashboard.py:1]: rm blank line

**Comment by ArthurJiang** [comment]
@micli @chaosddp please help review this pr.

**Review by Jinyu-W — APPROVED** [review]

File: docs/source/key_components/dashboard_visualization.rst [hunk 22]
```diff
@@ -0,0 +1,269 @@
...
+How to Use?
+-----------
+
+Generate dumped data
+~~~~~~~~~~~~~~~~~~~~
+
+The dumped data from environment is the data source of visualization.
+User needs to specify a parameter when creating Env object to determine
+if needs to dump data. To generate data, user needs to specify the
+parameter **options** with following code:
```
Comment (ArthurJiang) [inline docs/source/key_components/dashboard_visualization.rst:38]: To dump environment data for visualizing, user needs to ... when initialie the environment instance.

File: docs/source/key_components/dashboard_visualization.rst [hunk 23]
```diff
@@ -0,0 +1,269 @@
...
+to run experiment, and run this file with:
+
+.. code-block:: sh
+
+    python ./dump_data.py
+
+----
+
+User could specify dump destination folder by setting the parameter **options**.
+If user leave the value of **opts["enable-dump-snapshot"]** empty,
```
Comment (ArthurJiang) [inline docs/source/key_components/dashboard_visualization.rst:61]: rm "opts", User can set another dump target by ****

File: docs/source/key_components/dashboard_visualization.rst [hunk 24]
```diff
@@ -0,0 +1,230 @@
...
+        * cim
+        * citi_bike
+    topology:
+        * `cim <../scenarios/container_inventory_management.html#Topologies>`_
+
+        * `citi_bike <../scenarios/citi_bike.html#Topologies>`_
+    start_tick:
+        Integer not less than 0.
+    duration:
+        Integer not less than 1.
```
Comment (ArthurJiang) [inline docs/source/key_components/dashboard_visualization.rst:88]: rm

File: docs/source/key_components/dashboard_visualization.rst [hunk 25]
```diff
@@ -0,0 +1,230 @@
...
+.. code-block:: sh
+
+    python ./dump_data.py
+
+----
+
+Then data would be dumped automatically.
+
+The expected values of each parameter are listed below:
+
```
Comment (ArthurJiang) [inline docs/source/key_components/dashboard_visualization.rst:77]: rm

File: maro/data_lib/dump_csv_converter.py (hunk 7 above)
Comment (ArthurJiang) [reply maro/data_lib/dump_csv_converter.py:63]: @micli @Meroy9819  p1

**Pull request merged**